### DISCUSSION TITLE
Deployment in gh-pages cannot display 404 normally

### DISCUSSION THREAD

**Discussion by w1ndys** [squidfunk/mkdocs-material#7015] [Q&A] [answered]
The first picture is the wrong situation, and the second picture is the normal situation.
![Screenshot_2024-04-06-02-35-00-719_com android chrome](https://github.com/squidfunk/mkdocs-material/assets/109416673/1937aa5f-1dd6-4d96-a2b4-9d764bed41c0)
![Screenshot_2024-04-06-02-36-01-359_com android chrome](https://github.com/squidfunk/mkdocs-material/assets/109416673/0d5ccca8-2128-450f-ac69-bbf6473ee4ce)

**Comment by w1ndys** [comment]
And this is also true in zeabur deployment platform, and I am communicating with them.

**Comment by kamilkrzyskow** [comment]
Hello @W1ndys,
solely based on the shown URLs `/chinese/index.md` working and `/Easy-QFNU/chemistry.index.md` not working, I would assume you didn't set the `site_url` setting to the correct base URL:
https://www.mkdocs.org/user-guide/configuration/#site_url

Please keep in mind, to make sure everything works as expected, you would have to set it for each of your different deployment platforms that has a different base URL. You could set it with environmental variables: 
https://www.mkdocs.org/user-guide/configuration/#environment-variables

**Reply by w1ndys** [reply]
Hello, thank you for your help.，I set site_url as the address of gh-pages, which can solve the problem, but when I configure it through environment variables, the problem appears again. In addition, my deployment in zeabur also has this error, but no matter how the site_url is written, it will not solve the problem.

My profile:https://github.com/W1ndys/Easy-QFNU/blob/082f37784476c10ae834d06cd290a48710b768a7/mkdocs.yml#L8

![Screenshot_2024-04-07-01-15-41-639_com android chrome](https://github.com/squidfunk/mkdocs-material/assets/109416673/a9035f96-7807-4759-94e6-9088c647a86d)
![Screenshot_2024-04-07-01-15-14-954_com android chrome](https://github.com/squidfunk/mkdocs-material/assets/109416673/a964660c-ab34-4c6e-8775-4a402470f7c7)

**Reply by kamilkrzyskow** [reply, marked as answer]
On the screenshot it says that the variables are available in the `vars` context.
https://docs.github.com/en/actions/learn-github-actions/contexts#example-usage-of-the-vars-context

In the example it shows that you have to set the `env` with the `vars` variable:
```yaml
env:
  # Setting an environment variable with the value of a configuration variable
  SITE_URL: ${{ vars.SITE_URL }}
```
the `key` in the `env` corresponds to the name used in the `mkdocs.yml` file.

As for the other platform I don't know. You can debug it with a script.
The `!ENV` uses the provided key to access it in the `os.environ` object.
https://github.com/waylan/pyyaml-env-tag/blob/fdadeb7c859f43c6842bfb743943bfa77fa66f46/yaml_env_tag.py#L31

Therefore you could check with a script to see what variables are set in the deployment runner:
```py
import os

for pair in os.environ.items():
    print(pair)

name = "SITE_URL"
print(f"The value of {name} is", os.environ.get(name))
```

**Comment by w1ndys** [comment]
@kamilkrzyskow 
Hello
I tested and found some of my other deployment addresses
They are
https://ezqfnu.w1ndys.top/ (custom domain name in Vercel)
https://ezqfnu.pages.dev/  (the official domain name of cloudflare)
https://ezqfnu.vercel.app/  (the official domain name of Vercel)
https://ezqfnu.netlify.app/  (the official domain name of Netlify)
https://ezqfnu.zeabur.app/  (the official domain name of Zeabur)
https://w1ndys.github.io/Easy-QFNU/  (the official domain name of gh-pages)

My site_url is https://ezqfnu.w1ndys.top/
But I found that this 404 error occurs only in the deployment of Zeabur and gh-pages, and it is normal on other platforms (even if my site_url is not set to multiple).Just add `/chinese/index.md/` after the domain name to enter the page that detects whether it is a normal 404

So I guess the problem is not with site_url, maybe the problem is with the corresponding deployment platform?

normally page
![image](https://github.com/squidfunk/mkdocs-material/assets/109416673/bf70b316-d688-46d2-a964-3ba67b4b4ad4)

error page
![image](https://github.com/squidfunk/mkdocs-material/assets/109416673/dbee9d4f-cabc-4a17-98aa-c49828cf55a1)

**Reply by w1ndys** [reply]
In addition, the address in gh-pages should be related to site_URL, which has been tested, but Zeabur is not.

**Reply by kamilkrzyskow** [reply]
MkDocs builds a static website. Everything you see is a HTML file.

All other deployments work because they serve the /404.html file when an error occurs.

Not setting the site_url kind of works on most services because they use a "flat" base URL.

GitHub doesn't work because it uses the /Easy-QFNU/ project prefix.

As for Zeabur, the 404.html is present:
https://ezqfnu.zeabur.app/404.html
However, opening an invalid link doesn't serve it, but instead renders the home page (?) with relative file paths that are invalid on the invalid link.

The issue is on the Zeabur site, as it should serve the 404.html that contains absolute paths.

There was a similar issue with a  Nginx server:
- https://github.com/squidfunk/mkdocs-material/discussions/6685#discussioncomment-8246592

It had to be configured to serve the 404.html file.

**Reply by w1ndys** [reply]
Yes, I understand, thank you for your answer, I think the problem should appear on Zeabur
